File: cdm/java/mil/tatrc/physiology/datamodel/system/environment/actions/SEEnvironmentAction.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
package mil.tatrc.physiology.datamodel.system.environment.actions;

import org.jfree.util.Log;

import com.kitware.physiology.cdm.EnvironmentActions.EnvironmentActionData;
import com.kitware.physiology.cdm.EnvironmentActions.AnyEnvironmentActionData;

import mil.tatrc.physiology.datamodel.actions.SEAction;
import mil.tatrc.physiology.datamodel.substance.SESubstanceManager;

public abstract class SEEnvironmentAction extends SEAction
{
  public SEEnvironmentAction()
  {
    
  }
  
  public void copy(SEEnvironmentAction other)
  {
    if (this == other)
      return;
    super.copy(other);
  }
  
  public void reset()
  {
    super.reset();
  }
  
  public static void load(EnvironmentActionData src, SEEnvironmentAction dst) 
  {
    SEAction.load(src.getAction(), dst);
  }
  protected static void unload(SEEnvironmentAction src, EnvironmentActionData.Builder dst)
  {
    SEAction.unload(src, dst.getActionBuilder());
  }
  
  public static SEEnvironmentAction ANY2CDM(AnyEnvironmentActionData c, SESubstanceManager subMgr) 
  {
    switch(c.getActionCase())
    {
    case CONDITIONS:
    {
      SEChangeEnvironmentConditions dst = new SEChangeEnvironmentConditions();
      SEChangeEnvironmentConditions.load(c.getConditions(), dst, subMgr);
      return dst;
    }
    case THERMALAPPLICATION:
    {
      SEThermalApplication dst = new SEThermalApplication();
      SEThermalApplication.load(c.getThermalApplication(), dst);
      return dst;
    }
    case ACTION_NOT_SET:
      Log.warn("AnyEnvironmentActionData was empty...was that intended?");
      return null;
    }
    Log.error("Unsupported Environment Action type "+c.getActionCase());
    return null;
  }
  public static AnyEnvironmentActionData CDM2ANY(SEEnvironmentAction c)
  {
    AnyEnvironmentActionData.Builder dst = AnyEnvironmentActionData.newBuilder();
    
    if(c instanceof SEChangeEnvironmentConditions)
    {
      dst.setConditions(SEChangeEnvironmentConditions.unload((SEChangeEnvironmentConditions)c));
      return dst.build();
    }
    if(c instanceof SEThermalApplication)
    {
      dst.setThermalApplication(SEThermalApplication.unload((SEThermalApplication)c));
      return dst.build();
    }
    Log.error("Unsupported Environment Action type "+c);
    return dst.build();
  }
  
  public abstract String toString();
}

File: cdm/cpp/system/environment/actions/SEEnvironmentAction.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
#include "stdafx.h"
#include "system/environment/actions/SEEnvironmentAction.h"

SEEnvironmentAction::SEEnvironmentAction() : SEAction()
{

}

SEEnvironmentAction::~SEEnvironmentAction()
{
  Clear();
}

void SEEnvironmentAction::Clear()
{
  SEAction::Clear();
}

bool SEEnvironmentAction::IsValid() const
{
  return SEAction::IsValid();
}

void SEEnvironmentAction::Serialize(const cdm::EnvironmentActionData& src, SEEnvironmentAction& dst)
{
  SEAction::Serialize(src.action(), dst);
}

void SEEnvironmentAction::Serialize(const SEEnvironmentAction& src, cdm::EnvironmentActionData& dst)
{
  SEAction::Serialize(src, *dst.mutable_action());
}

#include "substance/SESubstanceManager.h"
#include "system/environment/actions/SEChangeEnvironmentConditions.h"
#include "system/environment/actions/SEThermalApplication.h"
SEEnvironmentAction* SEEnvironmentAction::Load(const cdm::AnyEnvironmentActionData& any, SESubstanceManager& subMgr)
{
  switch (any.Action_case())
  {
    case cdm::AnyEnvironmentActionData::ActionCase::kConditions:
    {
      SEChangeEnvironmentConditions* a = new SEChangeEnvironmentConditions(subMgr);
      SEChangeEnvironmentConditions::Load(any.conditions(), *a);
      return a;
    }
    case cdm::AnyEnvironmentActionData::ActionCase::kThermalApplication:
    {
      SEThermalApplication* a = new SEThermalApplication();
      SEThermalApplication::Load(any.thermalapplication(), *a);
      return a;
    }
  }
  subMgr.Error("Unknown action type : " + any.Action_case());
  return nullptr;
}
cdm::AnyEnvironmentActionData* SEEnvironmentAction::Unload(const SEEnvironmentAction& action)
{
  cdm::AnyEnvironmentActionData* any = new cdm::AnyEnvironmentActionData();
  const SEChangeEnvironmentConditions* cec = dynamic_cast<const SEChangeEnvironmentConditions*>(&action);
  if (cec != nullptr)
  {
    any->set_allocated_conditions(SEChangeEnvironmentConditions::Unload(*cec));
    return any;
  }
  const SEThermalApplication* ta = dynamic_cast<const SEThermalApplication*>(&action);
  if (ta != nullptr)
  {
    any->set_allocated_thermalapplication(SEThermalApplication::Unload(*ta));
    return any;
  }
  delete any;
  return nullptr;
}